components/brave_ads/browser/ads_tab_helper.cc

```diff
@@ -0,0 +1,177 @@
+#include "chrome/browser/ui/browser_list.h"
+#include "chrome/browser/ui/tabs/tab_strip_model.h"
+#include "components/dom_distiller/content/browser/distiller_page_web_contents.h"
+#include "components/dom_distiller/content/browser/web_contents_main_frame_observer.h"
+#include "components/dom_distiller/core/distiller_page.h"
+#include "components/dom_distiller/core/dom_distiller_service.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/render_frame_host.h"
+#include "content/public/browser/web_contents.h"
+#include "ui/base/resource/resource_bundle.h"
+
+namespace brave_ads {
+
+AdsTabHelper::AdsTabHelper(content::WebContents* web_contents)
+    : WebContentsObserver(web_contents),
+      tab_id_(SessionTabHelper::IdForTab(web_contents)),
+      ads_service_(nullptr),
+      weak_factory_(this) {
+  if (!tab_id_.is_valid())
+    return;
+
+  BrowserList::AddObserver(this);
+  Profile* profile = Profile::FromBrowserContext(
+      web_contents->GetBrowserContext());
+  ads_service_ = AdsServiceFactory::GetForProfile(profile);
+}
+
+AdsTabHelper::~AdsTabHelper() {
+  BrowserList::RemoveObserver(this);
+}
+
+void AdsTabHelper::DocumentOnLoadCompletedInMainFrame() {
+  // don't start distilling is the ad service isn't enabled
+  if (!ads_service_ || !ads_service_->is_enabled())
+    return;
+
+  auto* dom_distiller_service =
+      dom_distiller::DomDistillerServiceFactory::GetForBrowserContext(
+          web_contents()->GetBrowserContext());
+
+  if (!dom_distiller_service)
+    return;
+
+  auto source_page_handle =
+      std::make_unique<dom_distiller::SourcePageHandleWebContents>(
+            web_contents(), false);
+
+  auto distiller_page =
+      dom_distiller_service->CreateDefaultDistillerPageWithHandle(
+          std::move(source_page_handle));
+
+  auto options = dom_distiller::proto::DomDistillerOptions();
+  // options.set_extract_text_only(true);
+  // options.set_debug_level(1);
+
+  auto* distiller_page_ptr = distiller_page.get();
+
+  distiller_page_ptr->DistillPage(
+      web_contents()->GetLastCommittedURL(),
+      options,
+      base::Bind(&AdsTabHelper::OnWebContentsDistillationDone,
+          weak_factory_.GetWeakPtr(),
+          base::Passed(std::move(distiller_page))));
+}
+
+void AdsTabHelper::OnWebContentsDistillationDone(
+    std::unique_ptr<dom_distiller::DistillerPage> distiller_page,
+    std::unique_ptr<dom_distiller::proto::DomDistillerResult> distiller_result,
+    bool distillation_successful) {
+  if (!ads_service_ )
+    return;
+
+  if (distillation_successful &&
+      distiller_result->has_distilled_content() &&
+      distiller_result->has_markup_info() &&
+      distiller_result->distilled_content().has_html()) {
+    ads_service_->ClassifyPage(distiller_result->markup_info().url(),
+                               distiller_result->distilled_content().html());
+  } else {
+    // TODO(bridiver) - fall back to web_contents()->GenerateMHTML or ignore?
```

Comment:
Is this a question we need answering before 1.0?

Reply:
I don't think so